packages/remix-server-runtime/interface/cookies.ts

```diff
@@ -0,0 +1,65 @@
+import type { CookieParseOptions, CookieSerializeOptions } from "cookie";
+
+export interface CookieSignatureOptions {
+  /**
+   * An array of secrets that may be used to sign/unsign the value of a cookie.
+   *
+   * The array makes it easy to rotate secrets. New secrets should be added to
+   * the beginning of the array. `cookie.serialize()` will always use the first
+   * value in the array, but `cookie.parse()` may use any of them so that
+   * cookies that were signed with older secrets still work.
+   */
+  secrets?: string[];
+}
+  
+export type CookieOptions = CookieParseOptions &
+  CookieSerializeOptions &
+  CookieSignatureOptions;
+
+/**
+ * A HTTP cookie.
+ *
+ * A Cookie is a logical container for metadata about a HTTP cookie; its name
+ * and options. But it doesn't contain a value. Instead, it has `parse()` and
+ * `serialize()` methods that allow a single instance to be reused for
+ * parsing/encoding multiple different values.
+ *
+ * @see https://remix.run/api/remix#cookie-api
+ */
+export interface Cookie {
+  /**
+   * The name of the cookie, used in the `Cookie` and `Set-Cookie` headers.
+   */
+  readonly name: string;
+
+  /**
+   * True if this cookie uses one or more secrets for verification.
+   */
+  readonly isSigned: boolean;
+
+  /**
+   * The Date this cookie expires.
+   *
+   * Note: This is calculated at access time using `maxAge` when no `expires`
+   * option is provided to `createCookie()`.
+   */
+  readonly expires?: Date;
+
+  /**
+   * Parses a raw `Cookie` header and returns the value of this cookie or
+   * `null` if it's not present.
+   */
+  parse(
+    cookieHeader: string | null,
+    options?: CookieParseOptions
+  ): Promise<any>;
+
+  /**
+   * Serializes the given value to a string and returns the `Set-Cookie`
+   * header.
+   */
+  serialize(value: any, options?: CookieSerializeOptions): Promise<string>;
+}
+
+export type CreateCookie = (name: string, cookieOptions?: CookieOptions) => Cookie
```

Comment:
Just curious, any reason you prefer to use `type` over `interface` here? I typically prefer `interface` if I can get away with it, and use `type` when I need to create a type alias for a primitive or type union.

So e.g. I would've written this like this:

```ts
export interface CreateCookie {
  (name: string, options?: CookieOptions): Cookie;
}
```

Reply:
My rule-of-thumb is to use `interface` to define object shapes, and `type` for everything else. Also, since function types can only be applied to arrow functions (and not to `function`s), I like that the syntax mirrors its usage:

```ts
type CreateCookie = (name: string, cookieOptions?: CookieOptions) => Cookie 
const createCookie: CreateCookie = (name, cookieOptions) => { ... }
```

I'd use `interface` to define a function if I needed to attach additional properties to the function, but I don't think that's ever happened for me.